src/Emulator/Peripherals/Peripherals/MTD/OpenTitan_FlashController.cs

```diff
@@ -0,0 +1,1185 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using Antmicro.Renode.Logging;
+using Antmicro.Renode.Peripherals.Bus;
+using Antmicro.Renode.Core;
+using Antmicro.Renode.Core.Structure.Registers;
+using Antmicro.Renode.Peripherals.Memory;
+using Antmicro.Renode.Exceptions;
+using Antmicro.Renode.Utilities;
+
+namespace Antmicro.Renode.Peripherals.MTD
+{
+    public class OpenTitan_FlashController : IDoubleWordPeripheral, IKnownSize
+    {
+        enum ControlOp : uint
+        {
+            FLASH_READ = 0x0,
+            FLASH_PROGRAM = 0x1,
+            FLASH_ERASE = 0x2
+        }
+
+        enum ControlPartitionSel : uint
+        {
+            DATA_PARTITION = 0x0,
+            INFO_PARTITION = 0x1
+        }
+
+        enum ControlEraseSel : uint
+        {
+            PAGE_ERASE = 0x0,
+            BANK_ERASE = 0x1
+        }
+
+        enum OperationState{
+            INIT, WAITING, RUNNING, DONE
+        }
+
+        enum OperationType{
+            READ_DATA, PROGRAM_DATA, ERASE_DATA_PAGE
+        }
+
+        public OpenTitan_FlashController(Machine machine, MappedMemory flash)
+        {
+            this.Log(LogLevel.Info, "OpenTitan_FlashController/OpenTitan_FlashController: Entered");
+            var registersMap = new Dictionary<long, DoubleWordRegister>
+            {
+                // TODO(julianmb): support interrupts. no unittests exist though.
+                {(long)Registers.INTR_STATE, new DoubleWordRegister(this, 0x0)
+                    .WithFlag(0, name: "prog_empty", mode: FieldMode.Read | FieldMode.WriteOneToClear)
```

Comment:
The logic for these flags hasn't been implemented because the test suite I have doesn't exercise this aspect of the peripheral. I still wanted to keep the Read|WriteOneToClear behavior as that much is clearly laid out in the programmers guide.